purenix/src/Nix/Expr.hs

```diff
@@ -68,3 +81,29 @@ bin op a b = Expr $ Bin op a b
 
 path :: Text -> Expr
 path = Expr . Path
+
+numberedNames :: Text -> [Ident]
+numberedNames prefix = fmap (\n -> prefix <> T.pack (show n)) [0 :: Int ..]
+
+constructorFieldNames :: [Ident]
+constructorFieldNames = numberedNames "__field"
+
+negate :: Expr -> Expr
+negate = Expr . Not
+
+builtin :: Text -> Expr
+builtin = sel (var "builtins")
+
+--   Just
+-- becomes
+--   (a: { __tag = "Just"; __field0 = a; })
+constructor :: Ident -> [Ident] -> Expr
+constructor conName fields =
+  foldr
+    lam
+    ( attrs
+        fields
```

Comment:
I don't really feel strongly either way, but I guess you could take this out.

I imagine it might be confusing for users to have two ways to get to the field values.  

```suggestion
        []
```

Reply:
I wasn't sure what to best do here. My thinking was that it's nice to have named fields when using constructors in the FFI somewhere, but I guess that only works when deconstructing, since modifying/constructing requires properly handling the `__tag` and `__fieldX` attrs anyway.

The annoying thing is that, when pattern matching, `ConstructorBinder` doesn't give you the names of fields, just their positions, so you _need_ positional accessors.

Reply:
Oh, when I was playing around with this, I only saw the names being `value0`, `value1`, etc.  Are there times when these are named something different?

I was thinking that the field names will always be just `value0`, `value1`, etc, so there isn't much benefit to keeping them over the `field0` , `field1`, etc names that we are using.

Reply:
Did you try named record field accessors? I sort of assumed that the 'value' names were just defaults in case of nameless fields

Reply:
In PureScript, records work somewhat differently from Haskell.  I guess because they are "first-class" in PureScript.

So these two datatypes are basically the same in PureScript:

```purescript
data Foo1 = MkFoo1 { hello :: Int, bye :: String }

data Foo2 = MkFoo2 HelloRecord

type HelloRecord = { hello :: Int, bye :: String}
```

In this code, both `MkFoo1` and `MkFoo2` are basically the same.  They are both data constructors that take a single argument.  The argument they take is a record.

Here is the code that is generated for them from `purenix` currently:

```nix
let
  ...
  MkFoo2 = value0: 
    { inherit value0;
      __tag = "MkFoo2";
      __field0 = value0;
    };
  MkFoo1 = value0: 
    { inherit value0;
      __tag = "MkFoo1";
      __field0 = value0;
    };
  ...
```

I don't think there is any notion of named arguments for data constructors in PureScript.

-------------------------------------

In PureScript, you can also do things like having data constructors take two records:

```purescript
data Foo1 = MkFoo1 { hello :: Int, bye :: String } { dog :: Int, cat :: Number }
```

Reply:
Oh I see, that's really neat